src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -108,32 +129,13 @@ export class TableOfContents implements OnInit, AfterViewInit, OnDestroy {
     }
   }
 
-  private createLinks(): Link[] {
-    const links: Array<Link> = [];
-    const headers =
-        Array.from(this._document.querySelectorAll(this.headerSelectors)) as HTMLElement[];
-
-    if (headers.length) {
-      for (const header of headers) {
-        // remove the 'link' icon name from the inner text
-        const name = header.innerText.trim().replace(/^link/, '');
-        const {top} = header.getBoundingClientRect();
-        links.push({
-          name,
-          type: header.tagName.toLowerCase(),
-          top: top,
-          id: header.id,
-          active: false
-        });
-      }
-    }
-
-    return links;
-  }
-
   private onScroll(): void {
-    for (let i = 0; i < this.links.length; i++) {
-      this.links[i].active = this.isLinkActive(this.links[i], this.links[i + 1]);
+    const allLinks: Link[] = [];
+    this._linkSections.forEach(linkSection => {
+      allLinks.push(...linkSection.links);
```

Comment:
Why reconstruct this array on every scroll?

Reply:
Good catch, I refactored the function without looking at how often it's called. Changed so that the component keeps a cache on hand of the links it's displaying (new links appended when `addHeaders` is called, and reset in `resetHeaders`